### PR TITLE
[WIP] Added tags/favorites support

### PR DESCRIPTION
**Pull request by PVince81**
:construction: WIP

### CODE REVIEW

**Comment by PVince81** [comment]
@nickvergessen FYI

This is still WIP and doesn't work (JS loading issues, etc)

**Comment by PVince81** [comment]
The "favorites" list is now populated. It is **extremely** experimental and uses a _horribly wrong_\* approach. It seems we'll need to make many changes to core to support a proper approach.

@icewind1991 I'll need your support for this: https://github.com/owncloud/metadata/commit/3b864a1f2ab7d2e38bb1a485abc2f29310d711d3#diff-19747901543244a55807cbd25aac2737R71

Seems we'll need to add new API to be able to resolve fileids to proper Storage + FileInfo (FileInfo needs a Storage instance, which might be resolve-able based on the file id).

Also the file list will need to return the correct path because we have a flat list of files, but every file must know its path relative to the current user (for shared files)

File: appinfo/application.php [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+				$c->query('Request'),
+				$c->query('TagService')
+			);
+		});
+
+		/**
+		 * Core
+		 */
+		$container->registerService('UserId', function(IContainer $c) {
+			return \OCP\User::getUser();
```
Comment (LukasReschke) [inline appinfo/application.php:66]: Mhm. Why not inject IUser/IUserSession instead?

File: appinfo/application.php [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+         */
+        $container->registerService('L10N', function(IContainer $c) {
+            return $c->query('ServerContainer')->getL10N($c->query('AppName'));
+        });
+
+		/**
+		 * Services
+		 */
+		$container->registerService('Tagger', function(IContainer $c)  {
+			return \OC::$server->getTagManager()->load('files');
```
Comment (LukasReschke) [inline appinfo/application.php]: Query the server container instead?

File: appinfo/application.php [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+		});
+		$container->registerService('TagService', function(IContainer $c)  {
+			$homeFolder = \OC::$server->getUserFolder();
```
Comment (LukasReschke) [inline appinfo/application.php]: Query the server container instead?

File: controller/apicontroller.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		$this->tagService = $tagService;
+	}
+
+	/**
+	 * Updates the info of the specified file path
+	 * The passed tags are absolute, which means they will
+	 * replace the actual tag selection.
+	 *
+	 * @NoAdminRequired
+	 *
```
Comment (LukasReschke) [inline controller/apicontroller.php:35]: What about enabling CORS on that endpoint? (`@CORS`)

File: controller/apicontroller.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+		// if tags specified or empty array, update tags
+		if (!is_null($tags)) {
+			try {
+				$this->tagService->updateFileTags($path, $tags);
+			} catch (\OCP\Files\NotFoundException $e) {
+				return new JSONResponse($e->getMessage(), Http::STATUS_NOT_FOUND);
+			}
+			$result['tags'] = $tags;
+		}
+		return new JSONResponse($result, Http::STATUS_OK);
```
Comment (LukasReschke) [inline controller/apicontroller.php]: You can return a DataResponse instead and the AppFramework will use the correct response transformer.

File: controller/apicontroller.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+	 * @NoAdminRequired
+	 *
+	 * @param array $tagName tag name to filter by
+	 */
+	public function getFilesByTag($tagName) {
+		$result = array();
+		$files = $this->tagService->getFilesByTag($tagName);
+		$result['tag'] = $tagName;
+		$result['files'] = \OCA\Files\Helper::formatFileInfos($files);
+		return new JSONResponse($result, Http::STATUS_OK);
```
Comment (LukasReschke) [inline controller/apicontroller.php]: You can return a DataResponse instead and the AppFramework will use the correct response transformer.

File: appinfo/application.php (hunk 1 above)
Comment (PVince81) [inline appinfo/application.php:66]: I copied this from another app, so you might want to ask that person the same question :wink:

File: controller/apicontroller.php (hunk 6 above)
Comment (PVince81) [inline controller/apicontroller.php]: Nice, didn't know about this feature

File: appinfo/application.php (hunk 3 above)
Comment (nickvergessen) [inline appinfo/application.php]: Btw `$c` is `\OC::$server`, so you can just do `$c->...`

File: appinfo/application.php [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+	public function __construct (array $urlParams=array()) {
+		parent::__construct('metadata', $urlParams);
+
+		$container = $this->getContainer();
+
+        /**
+         * Core
+         */
+        $container->registerService('L10N', function(IContainer $c) {
```
Comment (nickvergessen) [inline appinfo/application.php]: Spaces are used for intendation?

File: appinfo/application.php (hunk 3 above)
Comment (PVince81) [inline appinfo/application.php]: Cool, thanks for the info.

**Comment by PVince81** [comment]
- [x] BUG: cannot tag on external storage...

**Comment by PVince81** [comment]
- [x] BUG: cannot favorite a shared folder or subfolder, only files

File: appinfo/application.php (hunk 2 above)
Comment (PVince81) [inline appinfo/application.php]: Doesn't work.
If I use:

```
return $c->query('TagManager')->load('files');
```

I get

```
Message: Identifier "TagManager" is not defined.
File: /srv/www/htdocs/owncloud/3rdparty/pimple/pimple/src/Pimple/Container.php
```

Comment (PVince81) [inline appinfo/application.php]: Never mind, wrong scope. Got it.

**Comment by PVince81** [comment]
- [x] BUG: creating a folder doesn't close "new" dropdown
- [x] BUG: favorite icon column too wide

**Comment by PVince81** [comment]
- [ ] test: sharing multiple times with same user, go to "Favorites" section
- [ ] TODO: delete/unshare hooks must delete tags

**Comment by PVince81** [comment]
Needs the following changes from core to work:
- [ ] Core changes: https://github.com/owncloud/core/pull/12360 (will be split into smaller PRs)
- [x] REQUIRES: getTagsForObjects: https://github.com/owncloud/core/pull/12409
- [ ] REQUIRES: plugin system for JS: https://github.com/owncloud/core/pull/12527
- [ ] REQUIRES: ~~extend node API with fileinfo: https://github.com/owncloud/core/pull/12614~~

**Comment by PVince81** [comment]
- [ ] REQUIRES: searchByTags: https://github.com/owncloud/core/pull/12617

**Comment by DeepDiver1975** [comment]
@PVince81 please use this travis.yaml - https://github.com/owncloud/files_antivirus/blob/master/.travis.yml - it will already setup core as required - including oracle - THX

**Comment by PVince81** [comment]
Ok thanks. Thing is, I used the `ocdev` tool and it auto-generated the Travis config.
Should I update ocdev then as well to use that ?

**Comment by DeepDiver1975** [comment]
> Should I update ocdev then as well to use that ?

that would be great - THX

**Comment by PVince81** [comment]
Obsolete. Favorites feature was moved to core https://github.com/owncloud/core/pull/12360 as suggested by @jancborchardt 

This means the "metadata" app will stay empty until we work on the next step which is providing tags support (tag management, etc) inside this separate app.

**Comment by PVince81** [comment]
Hmm thinking about it, let's keep it open but I'll strip out the favorites part.
We probably still want to keep the basic stuff and travis test case for the future.

**Comment by PVince81** [comment]
- [ ] TODO: strip down this PR to the minimum, groundwork for the future tags feature

**Pull request closed without merging**